### PR TITLE
test(mempool): impl update gas price threshold flow test

### PR DESCRIPTION
**Pull request by MohammadNassar1**
I've made `update_gas_price_threshold` API to be public, so I can use it flow test.

### CODE REVIEW

**Review by ayeletstarkware — COMMENTED** [review]
> Reviewed 2 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @MohammadNassar1)

___
*[`crates/mempool/tests/flow_test.rs` line 298 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBUnxEc1kFsFDkcTgBD:-OBUnxEd-VdbIlSRkHeR:b-u8uvfm) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L298)):*
> ```XML
>     // Setup.
>     let input_tip_100_gas_price_20 =
>         add_tx_input!(tx_hash: 1, address: "0x0", tip: 100, max_l2_gas_price: 20);
> ```

why do we need to mention the tip?



_Code quote:_
```XML
tip: 100
```

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware)

___
*[`crates/mempool/tests/flow_test.rs` line 298 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBUnxEc1kFsFDkcTgBD:-OBVRCjD-sTJklXsjdAT:bm7k663) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L298)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

why do we need to mention the tip?
</blockquote></details>

I added tip in the name as it's the priority order, and to show that although the first tx has a higher tip, mempool couldn't get the first tx as the gas price threshold is high.
wdyt?

**Review by ayeletstarkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @MohammadNassar1)

___
*[`crates/mempool/tests/flow_test.rs` line 298 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBUnxEc1kFsFDkcTgBD:-OBVlKbE49fGHNN6ZtQq:bbm5hkd) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L298)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

I added tip in the name as it's the priority order, and to show that although the first tx has a higher tip, mempool couldn't get the first tx as the gas price threshold is high.
wdyt?
</blockquote></details>

I think the tip should be tested in a unitest: create a mempool with high threshold, and one tx with high tip, low gas price, and call `get_txs`.

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @ayeletstarkware)

___
*[`crates/mempool/tests/flow_test.rs` line 298 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBUnxEc1kFsFDkcTgBD:-OBW1i9dCreVRPu4Cc7L:b-tsir6l) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L298)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

I think the tip should be tested in a unitest: create a mempool with high threshold, and one tx with high tip, low gas price, and call `get_txs`.
</blockquote></details>

I already have a unit test that checks get_txs functionality with different gas price values.

This test checks both `update_gas_price_threshold` and `get_txs`.

**Review by ayeletstarkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 4 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-:-OBZcDnJ2qIEKEsPjTTd:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @MohammadNassar1)

___
*[`crates/mempool/tests/flow_test.rs` line 298 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBUnxEc1kFsFDkcTgBD:-OBZaqNM0Dtn7gvvLNHh:b-dxoo6i) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L298)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

I already have a unit test that checks get_txs functionality with different gas price values.

This test checks both `update_gas_price_threshold` and `get_txs`.
</blockquote></details>

Ok, I'm not sure it's necessary since it's being checked in a unit test, but it's up to you.

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: all files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware)

___
*[`crates/mempool/tests/flow_test.rs` line 298 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBUnxEc1kFsFDkcTgBD:-OBZrngL6DC0wsFPcXC5:bt4n7tm) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L298)):*
<details><summary><i>Previously, ayeletstarkware (Ayelet Zilber) wrote…</i></summary><blockquote>

Ok, I'm not sure it's necessary since it's being checked in a unit test, but it's up to you.
</blockquote></details>

I've implemented a unit [test](https://github.com/starkware-libs/sequencer/blob/20845272bcb2c7c7bf9caa0347859febbecb2da1/crates/mempool/src/mempool_test.rs#L211-L222) that checks get_txs with pending queue.

We need this flow test, as it checks a flow that can happened and verify the invariant that mempool pops out only txs with gas prices higher than the threshold while updating the gas threshold.
___
*[`crates/mempool/tests/flow_test.rs` line 311 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBZsPNX1mCdT7b1I2rd:-OBZsPNY3dHVubbaeWR6:b4pddlr) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L311)):*
> ```XML
> 
>     mempool.update_gas_price_threshold(GasPrice(30));
>     get_txs_and_assert_expected(&mut mempool, 2, &[input_tip_50_gas_price_30.tx]);
> ```

In a real scenario, we would call `update_gas_price_threshold` only after a new round, which occurs after `commit_block`. 

I didn't add `commit_block` in this test to simplify it, as including it wouldn't affect the test outcome.

wdyt?



_Code quote:_
```XML
    mempool.update_gas_price_threshold(GasPrice(40));
    get_txs_and_assert_expected(&mut mempool, 2, &[]);

    mempool.update_gas_price_threshold(GasPrice(30));
    get_txs_and_assert_expected(&mut mempool, 2, &[input_tip_50_gas_price_30.tx]);
```

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: 0 of 8 files reviewed, 3 unresolved discussions (waiting on @ayeletstarkware and @elintul)

___
*[`crates/starknet_mempool/tests/flow_test.rs` line 307 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBeI1Ao0aoIeGOEZTeT:-OBeI1Ao0aoIeGOEZTeU:bohvbwj) ([raw file](https://github.com/starkware-libs/sequencer/blob/9fc55e9232b8ddb66438183aefadcc53cf290da2/crates/starknet_mempool/tests/flow_test.rs#L307)):*
> ```XML
>         add_tx(&mut mempool, input);
>     }
>     get_txs_and_assert_expected(&mut mempool, 2, &[input_gas_price_30.tx]);
> ```

This code checks the `AddTx` insertion to the correct queue.



_Code quote:_
```XML
    for input in [&input_gas_price_20, &input_gas_price_30] {
        add_tx(&mut mempool, input);
    }
    get_txs_and_assert_expected(&mut mempool, 2, &[input_gas_price_30.tx]);
```

**Review by elintul — APPROVED** [review]
> Reviewed 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: all files reviewed, 4 unresolved discussions (waiting on @ayeletstarkware and @MohammadNassar1)

___
*[`crates/mempool/tests/flow_test.rs` line 311 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBZsPNX1mCdT7b1I2rd:-OBeMp8O9hAumR6zv4kX:b-4q2ntt) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L311)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

In a real scenario, we would call `update_gas_price_threshold` only after a new round, which occurs after `commit_block`. 

I didn't add `commit_block` in this test to simplify it, as including it wouldn't affect the test outcome.

wdyt?
</blockquote></details>

I think we should add it in order for readers not to be confused, and test real scenarios. Please try to make it as lean as possible, while still sensible.
___
*[`crates/starknet_mempool/tests/flow_test.rs` line 295 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBeMDywDN3IaFHbthHz:-OBeMDywDN3IaFHbthI-:bsq1ghk) ([raw file](https://github.com/starkware-libs/sequencer/blob/9fc55e9232b8ddb66438183aefadcc53cf290da2/crates/starknet_mempool/tests/flow_test.rs#L295)):*
> ```XML
> 
> #[rstest]
> fn test_flow_update_gas_price_threshold(mut mempool: Mempool) {
> ```

Clear from the file's name.



_Suggestion:_
```XML
test_
```

File: crates/starknet_mempool/tests/flow_test.rs [hunk 1]
```diff
@@ -289,3 +290,26 @@ fn test_flow_commit_block_from_different_leader(mut mempool: Mempool) {
...
+    // Test: only txs with gas price above the threshold are returned.
+    mempool.update_gas_price_threshold(GasPrice(30));
+    for input in [&input_gas_price_20, &input_gas_price_30] {
+        add_tx(&mut mempool, input);
+    }
+    get_txs_and_assert_expected(&mut mempool, 2, &[input_gas_price_30.tx]);
+
+    let nonces = [("0x0", 1)];
+    let tx_hashes = [1];
+    commit_block(&mut mempool, nonces, tx_hashes);
```
[the remark this answers is not in the document]
Comment (MohammadNassar1) [reply crates/starknet_mempool/tests/flow_test.rs]: Done.

**Review by MohammadNassar1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: 7 of 8 files reviewed, 3 unresolved discussions (waiting on @ayeletstarkware, @elintul, and @graphite-app[bot])

___
*[`crates/mempool/tests/flow_test.rs` line 311 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-OBZsPNX1mCdT7b1I2rd:-OBehSen15tfjcgV4YYh:bb6un1j) ([raw file](https://github.com/starkware-libs/sequencer/blob/864866600c4f7a4d7e73d4ed3d2fc613ddab3be0/crates/mempool/tests/flow_test.rs#L311)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

I think we should add it in order for readers not to be confused, and test real scenarios. Please try to make it as lean as possible, while still sensible.
</blockquote></details>

Done

**Review by elintul — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882#-:-OBt9FcR0U58TS2xDsX8:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @ayeletstarkware and @graphite-app[bot])

**Review by MohammadNassar1 — COMMENTED** [review]
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/1882)* status: all files reviewed, 2 unresolved discussions (waiting on @ayeletstarkware, @elintul, and @graphite-app[bot])

**Pull request merged**